router2/src/dml_handlers/sharded_write_buffer.rs

```diff
@@ -162,30 +161,37 @@ where
 /// Enumerates all items in the iterator, maps each to a future that dispatches
 /// the [`DmlOperation`] to its paired [`Sequencer`], executes all the futures
 /// in parallel and gathers any errors.
-async fn parallel_enqueue<T>(v: T) -> Result<(), ShardError>
+///
+/// Returns a list of the sequences that were written
+async fn parallel_enqueue<T>(v: T) -> Result<Vec<DmlMeta>, ShardError>
 where
     T: Iterator<Item = (Arc<Sequencer>, DmlOperation)> + Send,
 {
-    let mut successes = 0;
-    let errs = v
-        .map(|(sequencer, op)| async move {
-            tokio::spawn(async move { sequencer.enqueue(op).await })
-                .await
-                .expect("shard enqueue panic")
-        })
-        .collect::<FuturesUnordered<_>>()
-        .filter_map(|v| {
-            if v.is_ok() {
-                successes += 1;
-            }
-            future::ready(v.err())
-        })
-        .collect::<Vec<WriteBufferError>>()
-        .await;
+    let mut successes = vec![];
+    let mut errs = vec![];
+
+    v.map(|(sequencer, op)| async move {
+        tokio::spawn(async move { sequencer.enqueue(op).await })
+            .await
+            .expect("shard enqueue panic")
+    })
+    // Use FuturesUnordered so the futures can run in parallel
+    .collect::<FuturesUnordered<_>>()
+    .collect::<Vec<_>>()
+    .await
+    // Sort the result into successes/failures upon completion
+    .into_iter()
+    .for_each(|v| match v {
+        Ok(meta) => successes.push(meta),
+        Err(e) => errs.push(e),
+    });
```

Comment:
Perhaps [`partition()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition) is a good candidate here:

```suggestion
    let (successes, errs): (Vec<_>, Vec<_>) = v
        .map(|(sequencer, op)| async move {
            tokio::spawn(async move { sequencer.enqueue(op).await })
                .await
                .expect("shard enqueue panic")
        })
        // Use FuturesUnordered so the futures can run in parallel
        .collect::<FuturesUnordered<_>>()
        .collect::<Vec<_>>()
        .await
        // Sort the result into successes/failures upon completion
        .into_iter()
        .partition(Result::is_ok);
```

Reply:
Thank you for the suggestion -- I actually tried `partition` in an earlier version of this code. The issue with using it is that then you end up with 2 `Vec<Result<WriteSummary>>` rather than a `Vec<WriteSummary>` and ` Vec<Error>`.

It is of course possible to then `unwrap` / `unwrap_err` each `Vec` , but I thought this way was a bit cleaner